Sources/SwiftFusion/Core/Manifold.swift

```diff
@@ -0,0 +1,110 @@
+/// A coordinate in a differentiable manifold's global coordinate system.
+///
+/// Includes a coordinate chart [1] between `Self` and `R^n`, centered at `self`.
+///
+/// [1] https://en.wikipedia.org/wiki/Topological_manifold#Coordinate_charts
+public protocol ManifoldCoordinate: Differentiable {
+  /// The local coordinate system used in the chart.
+  ///
+  /// Isomorphic to `R^n`, where `n` is the dimension of the manifold.
+  associatedtype LocalCoordinate: AdditiveArithmetic & Differentiable & VectorProtocol
+    where LocalCoordinate.TangentVector == LocalCoordinate
+
+  /// The global coordinate corresponding to `local` in the chart centered around `self`.
+  ///
+  /// Satisfies the following properties:
+  /// - `global(LocalCoordinate.zero) == self`
+  /// - There exists an open set `B` around `LocalCoordinate.zero` such that
+  ///   `local(global(b)) == b` for all `b \in B`.
+  @differentiable(wrt: local)
+  func global(_ local: LocalCoordinate) -> Self
+
+  /// The local coordinate corresponding to `global` in the chart centered around `self`.
+  ///
+  /// Satisfies the following properties:
+  /// - `local(self) == LocalCoordinate.zero`
+  /// - There exists an open set `B` around `self` such that `local(global(b)) == b` for all
+  ///   `b \in B`.
+  @differentiable(wrt: global)
+  func local(_ global: Self) -> LocalCoordinate
+}
+
+/// A point on a differentiable manifold.
+public protocol Manifold: Differentiable {
+  /// The manifold's global coordinate system.
+  associatedtype Coordinate: ManifoldCoordinate
+      where Coordinate.LocalCoordinate == Self.TangentVector
+
+  /// The coordinate of `self`.
+  ///
+  /// Note: This is not differentiable, and therefore clients should use `coordinate` instead.
+  var coordinateStorage: Coordinate { get set }
+
+  /// Creates a manifold point with coordinate `coordinateStorage`.
+  ///
+  /// Note: This is not differentiable, and therefore clients should use `init(coordinate:)`
+  /// instead.
+  init(coordinateStorage: Coordinate)
+}
+
+extension Manifold {
+  /// The coordinate of `self`.
+  @differentiable
+  public var coordinate: Coordinate { coordinateStorage }
```

Comment:
This swift syntax I still don't understand, even after looking for it in manual: what does the `{ coordinateStorage }` mean here?

Reply:
This means `coordinate` is a computed property that, when accessed, will return `coordinateStorage`

Reply:
Yeah, a recent swift feature is that single-expression function bodies implicitly return the value of the expression. I could start always using `return` if you think that would make things clearer.

Reply:
I knew about that feature. It would indeed, I think, make it more clear to the novel user here, as this is a double whammy otherwise: computed property and this new feature. For example, I did not understand that this could not be used to set. Adding `return` makes it clear that this is just a gettable property (even though storage is `get set`)